File: src/elixir/svd/lib/svd/exporters/base_exporter.ex
defmodule SVD.Exporters.BaseExporter do
  @moduledoc false

  alias SVD.Models
  alias SVD.Validators.ValidatorException

  @spec validate_and_export(
          module() | nil,
          term(),
          (struct() -> {:ok, term()} | {:error, term()})
        ) ::
          {:ok, term()} | {:error, term()}
  def validate_and_export(_validator_module, nil, _export_fun), do: {:error, "SVD cannot be nil"}

  def validate_and_export(validator_module, svd, export_fun) do
    case Models.normalize_dataset(svd) do
      nil ->
        {:error, "SVD cannot be nil"}

      dataset ->
        errors =
          if is_nil(validator_module) do
            []
          else
            validator_module.validate(dataset)
          end

        if errors == [] do
          export_fun.(dataset)
        else
          {:error, %ValidatorException{errors: errors}}
        end
    end
  end
end

File: src/go/svd/exporters/base_exporter.go
package exporters

import (
	"context"
	"errors"

	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/models"
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/validators"
)

// BaseExporter provides common functionality for all exporters.
type BaseExporter struct {
	validator validators.SVDValidator
}

// NewBaseExporter creates a new base exporter with the given validator.
func NewBaseExporter(validator validators.SVDValidator) *BaseExporter {
	return &BaseExporter{validator: validator}
}

// ValidateAndExport validates the SVD and calls the internal export method.
func (b *BaseExporter) ValidateAndExport(
	ctx context.Context,
	svd *models.StandardisedVesselDataset,
	exportFunc func(ctx context.Context, svd *models.StandardisedVesselDataset) (*StandardisedVesselDatasetContent, error),
) (*StandardisedVesselDatasetContent, error) {
	if ctx == nil {
		ctx = context.Background()
	}
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if svd == nil {
		return nil, errors.New("svd cannot be nil")
	}

	if b.validator != nil {
		validationErrors := b.validator.Validate(svd)
		if len(validationErrors) > 0 {
			return nil, validators.NewValidatorException(validationErrors)
		}
	}

	return exportFunc(ctx, svd)
}
